File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.cpp
// Last updated: 3/18/2026, 2:28:35 PM
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) return {1};

        int prefix = 1;

        vector<int> result(n, 1);

        for(int i = 0; i < n; i++) {
            result[i] = prefix;
            prefix *= nums[i];
        }

        int suffix = 1;

        for(int i = n - 1; i >= 0; i--) {
            result[i] *= suffix;
            suffix *= nums[i];
        }
        return result;
    }
};

File: 238-ProductofArrayExceptSelf/238-ProductofArrayExceptSelf.java
// Last updated: 8/8/2026, 11:21:40 AM
1class Solution {
2    public int[] productExceptSelf(int[] nums) {
3        int n = nums.length;
4        int[] ans = new int[n];
5
6        int left = 1;
7
8        for(int i = 0; i < n; i++) {
9            ans[i] = left;
10            left *= nums[i];
11        }
12        int right = 1;
13        for(int i = n - 1; i >= 0; i--) {
14            ans[i] *= right;
15            right *= nums[i];
16        }
17        return ans;
18    }
19}